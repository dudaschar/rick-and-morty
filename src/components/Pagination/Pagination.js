import React from 'react';
import PropTypes from 'prop-types';

import Button from 'src/components/Button/Button';

import { PaginationContainer } from './styles/Wrapper';
import { Text } from './styles/Typography';

function Pagination({ charactersInfo, onPrev, onNext }) {
  const prevURL = charactersInfo?.prev && new URL(charactersInfo.prev);
  const prevPage = prevURL?.searchParams.get('page')
  const currentPage = !prevURL ? 1 : Number(prevPage) + 1;
  
  return (
    <PaginationContainer>
      <Button
        disabled={!charactersInfo?.prev}
        onClick={onPrev}
      >
        Previous
      </Button>
      <Text>
        Page {currentPage} of {charactersInfo?.pages ? charactersInfo.pages : '1'}
      </Text>
      <Button
        disabled={!charactersInfo?.next}
        onClick={onNext}
      >
        Next
      </Button>
    </PaginationContainer>
  )
}

Pagination.defaultProps = {
  charactersInfo: {},
};

Pagination.propTypes = {
  charactersInfo: PropTypes.shape({
    count: PropTypes.number,
    next: PropTypes.string,
    pages: PropTypes.number,
    prev: PropTypes.string,
  }),
  onPrev: PropTypes.func.isRequired,
  onNext: PropTypes.func.isRequired,
}

export default Pagination;
