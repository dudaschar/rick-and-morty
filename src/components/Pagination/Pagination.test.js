import React from 'react';
import { fireEvent, render, screen } from 'src/testUtils/testUtils';

import Pagination from './Pagination';

describe('Pagination Component', () => {
  it('should render properly', () => {
    const onNextMock = jest.fn();
    const onPrevMock = jest.fn();
    const props = {
      count: 100,
      next: 'https://api.next.com',
      pages: 5,
      prev: 'https://api.prev.com',
    }

    const { asFragment } = render(
      <Pagination
        onNext={onNextMock}
        onPrev={onPrevMock}
        charactersInfo={props}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
  it('should call a function when clicking on control buttons', () => {
    const onNextMock = jest.fn();
    const onPrevMock = jest.fn();
    const props = {
      count: 100,
      next: 'https://api.next.com',
      pages: 5,
      prev: 'https://api.prev.com',
    }

    render(
      <Pagination
        onNext={onNextMock}
        onPrev={onPrevMock}
        charactersInfo={props}
      />
    );

    fireEvent.click(screen.getByText('Previous'));
    expect(onPrevMock).toHaveBeenCalledTimes(1);

    fireEvent.click(screen.getByText('Next'));
    expect(onNextMock).toHaveBeenCalledTimes(1);
  });
  it('should disable Previous button if it is the first page', () => {
    const onNextMock = jest.fn();
    const onPrevMock = jest.fn();
    const props = {
      count: 100,
      next: 'https://api.next.com',
      pages: 5,
      prev: null,
    };

    render(
      <Pagination
        onNext={onNextMock}
        onPrev={onPrevMock}
        charactersInfo={props}
      />
    );
    
    expect(screen.getByText('Previous').disabled).toBe(true);
    expect(screen.getByText('Next').disabled).toBe(false);
  });
  it('should disable Next button if it is the last page', () => {
    const onNextMock = jest.fn();
    const onPrevMock = jest.fn();
    const props = {
      count: 100,
      next: null,
      pages: 5,
      prev: 'https://api.prev.com',
    };

    render(
      <Pagination
        onNext={onNextMock}
        onPrev={onPrevMock}
        charactersInfo={props}
      />
    );
    
    expect(screen.getByText('Previous').disabled).toBe(false);
    expect(screen.getByText('Next').disabled).toBe(true);
  });
  it('should disable both buttons if it has only one page', () => {
    const onNextMock = jest.fn();
    const onPrevMock = jest.fn();
    const props = {
      count: 18,
      next: null,
      pages: 1,
      prev: null,
    };

    render(
      <Pagination
        onNext={onNextMock}
        onPrev={onPrevMock}
        charactersInfo={props}
      />
    );
    
    expect(screen.getByText('Previous').disabled).toBe(true);
    expect(screen.getByText('Next').disabled).toBe(true);
  });
  it('should enable both buttons if it has prev and next pages', () => {
    const onNextMock = jest.fn();
    const onPrevMock = jest.fn();
    const props = {
      count: 100,
      next: 'https://api.next.com',
      pages: 5,
      prev: 'https://api.prev.com',
    };

    render(
      <Pagination
        onNext={onNextMock}
        onPrev={onPrevMock}
        charactersInfo={props}
      />
    );
    
    expect(screen.getByText('Previous').disabled).toBe(false);
    expect(screen.getByText('Next').disabled).toBe(false);
  });
  it('should display correctly the current page', () => {
    const onNextMock = jest.fn();
    const onPrevMock = jest.fn();
    const props = {
      count: 100,
      next: 'https://api.next.com',
      pages: 5,
      prev: null,
    };

    const { rerender } = render(
      <Pagination
        onNext={onNextMock}
        onPrev={onPrevMock}
        charactersInfo={props}
      />
    );

    expect(screen.getByText('Page 1 of 5')).toBeTruthy();

    rerender(
      <Pagination
        onNext={onNextMock}
        onPrev={onPrevMock}
        charactersInfo={{
          count: 100,
          next: 'https://api.next.com/?page=3',
          pages: 5,
          prev: 'https://api.prev.com/?page=1'
        }}
      />
    );

    expect(screen.getByText('Page 2 of 5')).toBeTruthy();

    rerender(
      <Pagination
        onNext={onNextMock}
        onPrev={onPrevMock}
        charactersInfo={{
          count: 100,
          next: null,
          pages: 5,
          prev: 'https://api.prev.com/?page=4'
        }}
      />
    );

    expect(screen.getByText('Page 5 of 5')).toBeTruthy();
  });
});