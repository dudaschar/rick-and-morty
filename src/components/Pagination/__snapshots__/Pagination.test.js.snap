// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component should render properly 1`] = `
<DocumentFragment>
  <div
    class="Wrapper__PaginationContainer-sc-1u8z2uu-0 cTlRQS"
  >
    <button
      class="BaseButton-us6ig9-0 gNpRQk"
    >
      Previous
    </button>
    <span
      class="Typography__Text-sc-1byrzvq-0 dZlshT"
    >
      Page 1 of 5
    </span>
    <button
      class="BaseButton-us6ig9-0 gNpRQk"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;
